lib/pbench/server/api/resources/query_apis/__init__.py

```diff
@@ -224,24 +315,27 @@ def __init__(self, name: AnyStr, type: ParamType, required: bool = False):
         self.type = type
         self.required = required
 
-    def invalid(self, json: Dict[AnyStr, Any]) -> bool:
+    def invalid(self, json: JSON) -> bool:
         """
         Check whether the value of this parameter in the JSON document
         is invalid.
 
-        If the parameter is required and missing (or null), then it's invalid;
-        if the parameter is not required and specified as null, it's invalid.
-
         Args:
             json: The client JSON document being validated.
 
         Returns:
-            True if unacceptable
+            True if the specified value is unacceptable
         """
-        return not json[self.name] if self.name in json else self.required
+        if self.name in json:
+            return json[self.name] is None
+        else:
+            return self.required
```

Comment:
Again ...

```
try:
    return json[self.name] is None
except KeyError:
    return self.required
```

Reply:
Yes, I know that's a "Pythonic pattern" but I find intentionally planning exceptions to be extremely unnatural! I've done it in a few places, but I have to really think about it.

Reply:
I concur with Dave:  I think it's good and reasonable to use an EAFP approach when the failure is an exceptional case; however, when we're _expecting_ to fail, I think it's better to construct the logic as Dave has, here.

Reply:
The suggested code results in exactly 1 `name` lookup in `json` no matter the result.  The current code results in 2 `name` lookups whenever `name` is in `json`, and one when it is not.

This is a very low level routine, why pay the cost when you don't have to?

If it is a rare case for `name` to be in `json`, then we should have something like:

```
if self.name not in json:
    return self.required
return json[self.name] is None
```

To convey the intention and the leaning.  But that means we expect many "invalid" parameters, which is odd.

But happy to go with the consensus here.

Reply:
Yeah, I've (separately) come to the conclusion that the `in` operator is almost always undesirable, because, if the result is `True`, then the next thing to do is to fetch the value, so using `get()` and testing the result saves us from the second lookup.  However, in this particular case, having it not be `in` there is different from having it be in there but null/`None`.  So, using the `get()` approach is problematic (we would have to invent an "absent" value for use as the default value for `get()` to return, to avoid the ambiguity from a present-but-`None` value).

However, I don't think that the proposed code is so performance critical that we cannot afford two lookups.  And, the list that we're looking it up in is short, so the second one doesn't cost much anyway.

As for the two return statements, I would hesitate to draw any inference from their ordering (and, I wouldn't necessarily expect the compiler/interpreter to do so either).

Reply:
Not convinced.  But if the consensus is otherwise, so be it.

Reply:
We're validating user input, so "expected" is a loaded word. And while it makes sense to optimize for the success path, this isn't in an inner critical loop or anything so I doubt any performance difference is meaningful. (I'm not sure how smart Python's peephole analyzer is, but I'd take for granted that any compiled language would eliminate the second fetch anyway... which is probably part of my problem ;-) ) Mostly I find the pre-test logic easier to read; but that's mostly a matter of adjusting to a different style and it's not a real reason.